src/library/autodj/autodjprocessor.cpp

```diff
@@ -211,30 +217,75 @@ void AutoDJProcessor::fadeNow() {
-    } else if (rightDeck.isPlaying()) {
-        // Make sure rightDeck.fadeDuration is up to date.
-        calculateTransition(&rightDeck, &leftDeck);
-
-        // override posThreshold to start fade now
-        rightDeck.posThreshold = rightDeck.playPosition() -
-                ((1.0 - crossfader) / 2 * (rightDeck.fadeDuration));
-        // Repeat is disabled by FadeNow but disables auto Fade
-        rightDeck.setRepeat(false);
-    }
-    //else {
-    //    No deck playing, do not know what to do
+    DeckAttributes* pLeftDeck = m_decks[0];
+    DeckAttributes* pRightDeck = m_decks[1];
+    DeckAttributes* pFromDeck;
+    DeckAttributes* pToDeck;
+
+    if (pLeftDeck->isPlaying() &&
+            (!pRightDeck->isPlaying() || crossfader < 0.0)) {
+        pFromDeck = pLeftDeck;
+        pToDeck = pRightDeck;
+    } else if (pRightDeck->isPlaying()) {
+        pFromDeck = pRightDeck;
+        pToDeck = pLeftDeck;
+    } else {
+        // Both decks are playing or no decks are playing;
+        // fading now makes no sense.
+        return;
+    }
+
+    pFromDeck->setRepeat(false);
+    pFromDeck->isFromDeck = true;
+    pToDeck->isFromDeck = false;
+
+    double fromDeckCurrentPosition = pFromDeck->playPosition() * pFromDeck->duration();
+    double toDeckCurrentPosition = pToDeck->playPosition() * pToDeck->duration();
+
+    pFromDeck->fadeBeginPos = fromDeckCurrentPosition;
+    // Do not seek to a calculated start point; start the to deck from wherever
+    // it is if the user has seeked since loading the track.
+    pToDeck->startPos = toDeckCurrentPosition;
+
+    // If the user presses "Fade now", assume they want to fade *now*, not later.
+    // So if the spinbox time is negative, do not insert silence.
+    double spinboxTime = fabs(m_transitionTime);
+
+    double fadeTime;
+    if (m_transitionMode == TransitionMode::FullIntroOutro
+       || m_transitionMode == TransitionMode::AlignIntroOutroStart) {
+        // Use the intro length as the transition time. If the user has seeked
+        // away from the intro start since the track was loaded, start from
+        // there and do not seek back to the intro start. If they have seeked
+        // past the introEnd or the introEnd is not marked, fall back to the
+        // spinbox time.
+        double outroEnd = getOutroEndPosition(pFromDeck);
+        double introEnd = getIntroEndPosition(pToDeck);
+
+        if (toDeckCurrentPosition <= introEnd && introEnd >= 0) {
+            double introLength = introEnd - toDeckCurrentPosition;
+            if (outroEnd >= 0) {
+                // The fade must end by the outro end at the latest.
+                double timeUntilOutroEnd = outroEnd - fromDeckCurrentPosition;
+                fadeTime = math_min(introLength, timeUntilOutroEnd);
+            } else {
+                fadeTime = introLength;
+            }
+        } else {
+            fadeTime = spinboxTime;
+        }
+    } else {
+        fadeTime = spinboxTime;
```

Comment:
In case of skip silence, fadeTime should be limited by the silence start point.

Reply:
This is not necessary. The time between the silence start point and the playposition will always be greater than the spinbox time, otherwise the transition will have already started without pressing "Fade now".

Reply:
This happens if the track is shorter than the transition time or if auto DJ is started within the outro.

Reply:
If the track is shorter than transition time, won't fading have already started?

> if auto DJ is started within the outro

What do you mean? The outro is not taken into account in the skip silence mode.

Reply:
The outro end is already not taken into account for the skip silence and full track modes. There is nothing that needs to change here.

> It would be just balanced to have a skip silence branch in a same way we have Intro/Outro branch.

This is unnecessary. The requirements of the skip silence and full track mode are the same and are both covered by the else branch. If you want I can explicitly add `else if (m_transitionMode == TransitionMode::FixedFullTrack || m_transitionMode == TransitionMode::FixedSkipSilence)` instead of just `else`, but that's also not necessary.

Reply:
If you insist, but all you would be doing is overcomplicating the code.